server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceShardsAllocator.java

```diff
@@ -193,6 +200,28 @@ public void allocate(RoutingAllocation allocation, ActionListener<Void> listener
         reconcile(currentDesiredBalance, allocation);
     }
 
+    private void processNodeShutdowns(ClusterState clusterState) {
+        final var nodes = clusterState.nodes();
+        final var nodeShutdowns = clusterState.metadata().nodeShutdowns();
+        // If we remove a shutdown marker from a node, but it is still in the cluster, we'd need a reset.
+        boolean reset = processedNodeShutdowns.stream()
+            .anyMatch(nodeId -> nodeShutdowns.contains(nodeId) == false && nodes.get(nodeId) != null);
+        // Clean up processed shutdowns that are removed from the cluster metadata
+        processedNodeShutdowns.removeIf(nodeId -> nodeShutdowns.contains(nodeId) == false);
```

Comment:
i could merge these two loops, but i find it more readable this way (and the set size should be always small).

Reply:
NIT: could be simplified with ` processedNodeShutdowns.retainAll(nodeShutdowns.keySet());`